### PR TITLE
Better reporting on payment failures

### PR DESCRIPTION
**Pull request by kmmatwork**

### CODE REVIEW

File: prime-modules/src/main/kotlin/org/ostelco/prime/paymentprocessor/core/PaymentError.kt [hunk 1]
```diff
@@ -2,14 +2,99 @@ package org.ostelco.prime.paymentprocessor.core
...
+ * @param code A short string indicating the type of error from upstream
+ * @param declineCode A short string indication the reason for a card
+ *                    error from the card issuer
+ * @param internalError Internal error chaining
+ */
+sealed class PaymentError(val description: String,
+                          val message: String? = null,
+                          val code: String? = null,
+                          val declineCode: String? = null,
+                          val internalError: InternalError?) : InternalError()
```
Comment (vihangpatil) [inline prime-modules/src/main/kotlin/org/ostelco/prime/paymentprocessor/core/PaymentError.kt:30]: This is really good.  Mapping out all the Stripe errors! 👍

File: neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt [hunk 2]
```diff
@@ -1596,20 +1606,24 @@ object Neo4jStoreSingleton : GraphStore {
                         it
                     }.linkReversalActionToTransaction(transaction) {
                         paymentProcessor.removeInvoice(it.id)
-                        logger.error(NOTIFY_OPS_MARKER,
-                                """Failed to create or pay invoice for customer ${customer.id}, invoice-id: ${it.id}.
+                        logger.warn(NOTIFY_OPS_MARKER,
+                                """Failed to pay invoice for customer ${customer.id}, invoice-id: ${it.id}.
                                    Verify that the invoice has been deleted or voided in Stripe dashboard.
                                 """.trimIndent())
```
Comment (vihangpatil) [inline neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt]: Move the starting `"""` to the line above.  Keeping it in the same line prevents trimIndent() from working.

Reference: https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.text/trim-indent.html

File: neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt [hunk 3]
```diff
@@ -1596,20 +1606,24 @@ object Neo4jStoreSingleton : GraphStore {
...
+                                message = "Failed to complete purchase of product $sku for $price.amount} ${price.currency} " +
+                                        "status: ${it.code} decline reason: ${it.declineCode}")
                         it
                     }.linkReversalActionToTransaction(transaction) {
                         paymentProcessor.refundCharge(it.chargeId)
-                        logger.error(NOTIFY_OPS_MARKER,
+                        logger.warn(NOTIFY_OPS_MARKER,
                                 """Refunded customer ${customer.id} for invoice: ${it.id}.
                                    Verify that the invoice has been refunded in Stripe dashboard.
                                 """.trimIndent())
```
Comment (vihangpatil) [inline neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt]: Move the starting `"""` to the line above.  Keeping it in the same line prevents trimIndent() from working.

Reference: https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.text/trim-indent.html

File: neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt [hunk 4]
```diff
@@ -1317,12 +1319,17 @@ object Neo4jStoreSingleton : GraphStore {
...
                             .mapLeft {
-                                logger.error("Failed to save purchase record for customer ${customer.id}, invoice-id $invoiceId, invoice will be voided in Stripe")
-                                AuditLog.error(customerId = customer.id, message = "Failed to save purchase record - invoice-id $invoiceId, invoice will be voided in Stripe")
-                                BadGatewayError("Failed to save purchase record",
-                                        error = it)
+                                logger.error("Failed to save purchase record for customer ${customer.id}, invoice: $invoiceId, invoice will be voided in Stripe")
+                                AuditLog.error(customerId = customer.id,
+                                        message = "Failed to save purchase record - invoice: $invoiceId, invoice will be voided in Stripe")
+                                StorePurchaseError("Failed to save purchase record",
+                                        internalError = it)
```
Comment (vihangpatil) [inline neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt:1326]: This is good.  Better Error type.

File: neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt [hunk 5]
```diff
@@ -1317,12 +1319,17 @@ object Neo4jStoreSingleton : GraphStore {
...
                             }.bind()
 
+                    /* Adds purchase to customer history. */
+                    AuditLog.info(customerId = customer.id,
+                            message = "Purchased product $sku for ${product.price.amount} ${product.price.currency} (invoice: $invoiceId, charge-id: $chargeId)")
```
Comment (vihangpatil) [inline neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt]: `product.price.amount` is in lowest denomination (cents / øres). 
Do floating divide by 100.0 and then print up to 2 decimal places.

File: neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt [hunk 6]
```diff
@@ -1520,7 +1529,7 @@ object Neo4jStoreSingleton : GraphStore {
                     }
                     PaymentStatus.REQUIRES_PAYMENT_METHOD -> {
                         NotCreatedError(type = planEntity.name, id = "Failed to subscribe $customerId to ${plan.id}",
```
Comment (vihangpatil) [inline neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt]: NotCreatedError should have `type = "Customer subscription to Plan"` and `id = "$customerId -> ${plan.id}"`.

File: neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt [hunk 7]
```diff
@@ -1596,20 +1606,24 @@ object Neo4jStoreSingleton : GraphStore {
...
                     }.bind()
 
             /* Force immediate payment of the invoice. */
             val invoicePaymentInfo = paymentProcessor.payInvoice(invoice.id)
                     .mapLeft {
-                        logger.error("Payment of invoice ${invoice.id} failed for customer ${customer.id}.")
+                        logger.warn("Payment of invoice ${invoice.id} failed for customer ${customer.id}.")
+                        /* Adds failed purchase to customer history. */
+                        AuditLog.warn(customerId = customer.id,
+                                message = "Failed to complete purchase of product $sku for $price.amount} ${price.currency} " +
```
Comment (vihangpatil) [inline neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt]: `$price.amount}` printed as string literal due to missing `{`.  Also, it prints øres / cents.  Divide by 100.0 and print up to 2 decimal places.

File: neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt (hunk 3 above)
Comment (kmmatwork) [reply neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt]: Fixed in update.

File: neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt (hunk 6 above)
Comment (kmmatwork) [reply neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt]: Fixed.

File: neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt (hunk 7 above)
Comment (kmmatwork) [reply neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt]: Fixed.

**Review by vihangpatil — APPROVED** [review]
Thanks @kmmatwork

**Pull request merged**